### ISSUE TITLE
Change virtualization options for nixos-rebuild build-vm 

### ISSUE THREAD

**Issue by Enteee** [NixOS/nixpkgs#59219]
## Issue description

I would like to build and run a qemu vm from my current nixos configuration using `nixos-rebuild build-vm`. The problem is that the generated run-script does not start the vm with a configuration which allows to run hardware demanding applications in it. For this reason I would like to tweak some options which are exposed by the [qemu-vm module](https://github.com/NixOS/nixpkgs/blob/c51fcbc82f0ff1268e67a5fb517baacad969ba0d/nixos/modules/virtualisation/qemu-vm.nix#L195). But when I do this `nixos-rebuild` raises errors. 

### Steps to reproduce

Add:

```
virtualization.cores = 5;
```
to `/etc/nixos/configuration.nix`.  And run 

```
$ nixos-rebuild build-vm 
building Nix...
error: The option `virtualisation.cores' defined in `/etc/nixos/configuration.nix' does not exist.
(use '--show-trace' to show detailed location information)
building the system configuration...
these derivations will be built:
```

Interestingly the run script was built just fine. 

Also when I run  `nixos-rebuild build-vm --fast` the error disappears.

Is this a bug or how can I define `virtualisation.cores` without raising this error on `nixos-rebuild` ?

## Technical details

 - system: `"x86_64-linux"`
 - host os: `Linux 4.14.110, NixOS, 19.09pre174426.acbdaa569f4 (Loris)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.2`
 - channels(root): `"nixos-19.09pre174426.acbdaa569f4, nixos-hardware, nixpkgs-19.09pre174594.0c0954781e2"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

**Comment by Enteee** [comment]
My current work-around for this issue is defining a system environment variable `QEMU_OPTS`.
```
environment.variables = {
    QEMU_OPTS = "-m 4096 -smp 4 -enable-kvm";
 };
```
Nevertheless I would still like to use the `virtualization.*` options.

**Comment by peterhoeg** [comment]
Still an issue.

**Comment by bobvanderlinden** [comment]
Maybe not ideal, but it seems to be working to do the following in configuration.nix:

```
{ config, pkgs, ...}:
{
  imports = [ <nixpkgs/nixos/modules/virtualisation/qemu-vm.nix> ];
  virtualisation.memorySize = "1G";
}
```

**Comment by kirillrdy** [comment]
I have ran into same issue, but @bobvanderlinden solutions works for me

**Comment by chasecaleb** [comment]
I think this can be closed between #151082 and https://github.com/NixOS/nixpkgs/issues/59219#issuecomment-774711048, right?

**Comment by ParetoOptimalDev** [comment]
I just ran into this with a recent unstable and it's quite annoying.

> I think this can be closed between https://github.com/NixOS/nixpkgs/pull/151082 and https://github.com/NixOS/nixpkgs/issues/59219#issuecomment-774711048, right?

Well https://github.com/NixOS/nixpkgs/issues/59219#issuecomment-774711048 is just a workaround so I'd say no.

As for https://github.com/NixOS/nixpkgs/pull/151082 neither of these options work, but perhaps I'm trying to use it wrong:

```
virtualisation.vmVariant.memorySize = 4906;
virtualisation.qemu-vm.memorySize = 4906;
virtualisation.qemu.memorySize = 4906;
```

So I'd say it shouldn't be closed yet.

**Comment by Radvendii** [comment]
I agree that this is a workaround and not a good final solution, but if you want to do it in pure evaluation mode, this should work instead
```
{ modulesPath, ...}:
{
  imports = [ (modulesPath + "/virtualisation/qemu-vm.nix") ];
  virtualisation.memorySize = "1G";
}
```

**Comment by NiklasGollenstede** [comment]
These options didn't exist when https://github.com/NixOS/nixpkgs/issues/59219#issuecomment-774711048 was written, but I'd say the way to do this now (which also doesn't mess with your non-vm builds of the system) is via `virtualisation.vmVariant`(`WithBootLoader`). E.g.:
```nix
{ virtualisation.vmVariant.virtualisation.cores = 5;
  # and/or
  virtualisation.vmVariantWithBootLoader.virtualisation.cores = 5;
```
You can also read the VM version(s) of the configuration via `config.virtualisation.vmVariant`. E.g.:
```nix
{ virtualisation.vmVariant.boot.resumeDevice = lib.mkIf config.virtualisation.vmVariant.virtualisation.useDefaultFilesystems (lib.mkVMOverride "");
```

It's a bit verbose and you may have to repeat everything for `vmVariant` and `vmVariantWithBootLoader`, but it works.

**Comment by NiklasGollenstede** [comment]
Thinking about it a bit more, what is really wrong here are the options search and the manual.

Since `nixos/modules/virtualisation/qemu-vm.nix` is not imported by default (and I guess this fact is not known to either the search or the manual), `virtualisation.{fileSystems,memorySize,msize,diskSize,diskImage,bootDevice,emptyDiskImages,graphics,resolution,cores,sharedDirectories,additionalPaths,forwardPorts,vlans,writableStore,writableStoreUseTmpfs,host,qemu,useNixStoreImage,useBootLoader,useEFIBoot,efi,useDefaultFilesystems,efiVars,bios}` are not available as an option.

But `nixos/modules/bootloader/extlinux.nix.md` imports that file into the `virtualisation.vmVariant`(`WithBootLoader`) configs. The `virtualisation.{...}` options are available as sub-options here, but are hidden by the `mkOption { visible = "shallow"; }` (cuz otherwise the whole option tree would be duplicated here?).

IMHO it actually makes quite a lot of sense the way that it _works_, but the documentation is wrong, and it would be bad if it was correct (all those options should not be bunched in with other virtualization technologies in `virtualisation.*`).
Also, maybe `virtualisation.{...}` should be renamed to `vm.{...}` (or something), to make e.g. `virtualisation.vmVariant.virtualisation.cores` a bit less redundant/confusing/verbose (as `virtualisation.vmVariant.vm.cores`).
